src/feature/hs/hs_service.c

```diff
@@ -1686,6 +1686,15 @@ build_desc_intro_points(const hs_service_t *service,
 
   DIGEST256MAP_FOREACH(desc->intro_points.map, key,
                        const hs_service_intro_point_t *, ip) {
+    if (!ip->circuit_established) {
+      /* Ignore un-established intro points. They can linger in that list
+       * because their circuit has not opened and they haven't been removed
+       * yet even though we have enough intro circuits.
+       *
+       * XXX: Due to #31561, it can stay in that list until rotation so this
```

Comment:
I don't think it makes sense to have this alarming XXX in master where we will also merge #31561. Should we have two branches, one with the XXX for maint that don't include #31561 and one without for master? Or maybe just remove the XXX entirely? Up to you.